### PR TITLE
add more tests for auto reexecution behavior

### PR DESCRIPTION
**Pull request by jamiedemaria**
## Summary & Motivation
Adds some additional test cases to our auto re-execution test suite to test for things like daemon failures at specific points in the iteration

### CODE REVIEW

**Comment by jamiedemaria** [comment]
* **#26277** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26277?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26277?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `master`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: integration_tests/test_suites/daemon-test-suite/auto_run_reexecution_tests/test_auto_run_reexecution.py [hunk 1]
```diff
@@ -422,6 +425,226 @@ def test_consume_new_runs_for_automatic_reexecution(instance, workspace_context)
...
+        consume_new_runs_for_automatic_reexecution(
+            workspace_context,
+            instance.get_run_records(filters=RunsFilter(statuses=[DagsterRunStatus.FAILURE])),
+        )
+    )
+
+    # because the retried run already exists, it is in the run group for the original run. Because
+    # it is in the run group of the original run, we will see that the run group contains a run with
+    # parent_run_id as the original run. Therefore the original run will not be retried
+    # This is a problem, since the retried run will be stuck in a NOT_STARTED state
```
Comment (gibsondan) [inline integration_tests/test_suites/daemon-test-suite/auto_run_reexecution_tests/test_auto_run_reexecution.py:487]: This edge case exists in master right? Curious if your subsequent changes are going to detect this case and re-enqueue it

File: integration_tests/test_suites/daemon-test-suite/auto_run_reexecution_tests/test_auto_run_reexecution.py [hunk 2]
```diff
@@ -422,6 +425,226 @@ def test_consume_new_runs_for_automatic_reexecution(instance, workspace_context)
...
+    assert len(instance.get_runs()) == 1  # just the original run
+
+    list(
+        consume_new_runs_for_automatic_reexecution(
+            workspace_context,
+            instance.get_run_records(filters=RunsFilter(statuses=[DagsterRunStatus.FAILURE])),
+        )
+    )
+
+    assert len(instance.run_coordinator.queue()) == 1
```
Comment (gibsondan) [inline integration_tests/test_suites/daemon-test-suite/auto_run_reexecution_tests/test_auto_run_reexecution.py:546]: this one on the other hand you are definitely not going to fix right? (which i think is fine, this is a pretty unlikely edge case where you have to delete the run at exactly the wrong time)

maybe worth a comment explaining that this is a known race condition where things might not behave as you might expect

**Review by gibsondan — APPROVED** [review]
nice - thanks for adding these. The other pattern I have seen us use for this is adding a 'debug_crash_flags' dictionary that the machinery checks at various points and raises an exception (or even crashes) if the flags are set, but I think this works too.

some of these are kind of verifying the exact behavior of certain edge cases that don't necessarily end up in a 100% expected state and it might be good to make sure that is clear in comments in the various tests, so that subseqeuent behavior changes know how bad it is if the test starts failing, but because we have actually fixed an edge case and made it better

File: integration_tests/test_suites/daemon-test-suite/auto_run_reexecution_tests/test_auto_run_reexecution.py (hunk 1 above)
Comment (jamiedemaria) [reply integration_tests/test_suites/daemon-test-suite/auto_run_reexecution_tests/test_auto_run_reexecution.py:487]: yeah, the current behavior in master is that the run won't get resubmitted. The stacked PR can detect this case and resubmit

File: integration_tests/test_suites/daemon-test-suite/auto_run_reexecution_tests/test_auto_run_reexecution.py (hunk 2 above)
Comment (jamiedemaria) [reply integration_tests/test_suites/daemon-test-suite/auto_run_reexecution_tests/test_auto_run_reexecution.py:546]: added a comment, but you're correct, i'm not changing this behavior.
